src/modules/mc_rate_control/MulticopterRateControl.cpp

```diff
@@ -0,0 +1,542 @@
+		_v_control_mode_sub.update(&_v_control_mode);
+		_vehicle_land_detected_sub.update(&_vehicle_land_detected);
+
+		vehicle_status_poll();
+		vehicle_motor_limits_poll();
+
+		const bool manual_control_updated = _manual_control_sp_sub.update(&_manual_control_sp);
+
+		// generate the rate setpoint from sticks?
+		bool manual_rate_sp = false;
+
+		if (_v_control_mode.flag_control_manual_enabled &&
+		    !_v_control_mode.flag_control_altitude_enabled &&
+		    !_v_control_mode.flag_control_velocity_enabled &&
+		    !_v_control_mode.flag_control_position_enabled) {
+
+			// landing gear controlled from stick inputs if we are in Manual/Stabilized mode
+			//  limit landing gear update rate to 50 Hz
+			if (hrt_elapsed_time(&_landing_gear.timestamp) > 20_ms) {
+				_landing_gear.landing_gear = get_landing_gear_state();
+				_landing_gear.timestamp = hrt_absolute_time();
+				_landing_gear_pub.publish(_landing_gear);
+			}
+
+			if (!_v_control_mode.flag_control_attitude_enabled) {
+				manual_rate_sp = true;
+			}
+
+			// Check if we are in rattitude mode and the pilot is within the center threshold on pitch and roll
+			//  if true then use published rate setpoint, otherwise generate from manual_control_setpoint (like acro)
+			if (_v_control_mode.flag_control_rattitude_enabled) {
+				manual_rate_sp =
+					(fabsf(_manual_control_sp.y) > _param_mc_ratt_th.get()) ||
+					(fabsf(_manual_control_sp.x) > _param_mc_ratt_th.get());
+			}
+
+		} else {
+			_landing_gear_sub.update(&_landing_gear);
+		}
+
+		const bool is_hovering = (_vehicle_status.vehicle_type == vehicle_status_s::VEHICLE_TYPE_ROTARY_WING)
+					 && !_vehicle_status.in_transition_mode;
+
+		if (manual_rate_sp && is_hovering) {
+			if (manual_control_updated) {
+				// manual rates control - ACRO mode
+				Vector3f man_rate_sp(
+					math::superexpo(_manual_control_sp.y, _param_mc_acro_expo.get(), _param_mc_acro_supexpo.get()),
+					math::superexpo(-_manual_control_sp.x, _param_mc_acro_expo.get(), _param_mc_acro_supexpo.get()),
+					math::superexpo(_manual_control_sp.r, _param_mc_acro_expo_y.get(), _param_mc_acro_supexpoy.get()));
+
+				_rates_sp = man_rate_sp.emult(_acro_rate_max);
+				_thrust_sp = _manual_control_sp.z;
+
+				// publish rate setpoint
+				vehicle_rates_setpoint_s v_rates_sp{};
+				v_rates_sp.roll = _rates_sp(0);
+				v_rates_sp.pitch = _rates_sp(1);
+				v_rates_sp.yaw = _rates_sp(2);
+				v_rates_sp.thrust_body[0] = 0.0f;
+				v_rates_sp.thrust_body[1] = 0.0f;
+				v_rates_sp.thrust_body[2] = -_thrust_sp;
+				v_rates_sp.timestamp = hrt_absolute_time();
+
+				_v_rates_sp_pub.publish(v_rates_sp);
+			}
+
+		} else {
+			// use rates setpoint topic
+			vehicle_rates_setpoint_s v_rates_sp;
+
+			if (_v_rates_sp_sub.update(&v_rates_sp)) {
+				_rates_sp(0) = v_rates_sp.roll;
+				_rates_sp(1) = v_rates_sp.pitch;
+				_rates_sp(2) = v_rates_sp.yaw;
+				_thrust_sp = -v_rates_sp.thrust_body[2];
+
+			}
+		}
+
+		// update scaled rates (TPA) on throttle change
```

Comment:
Nice optimisation, but if the thrust setpoint never changes the gains are never initialized (see the comment about `parameters_updated()` above)

Reply:
Yes this was a bit of a work in progress. 

TPA (with default disabled parameters) was more than half a percent of total cpu usage.